File: services/mod-rev-core/internal/observability/observability.go
// Package observability provides zero-dependency Prometheus-style metrics
// for the Go services (Stage 7.C): a stdlib-only counter/histogram registry,
// HTTP middleware, and a /metrics text-exposition handler. Equivalent to
// services/_shared/observability.py used by the FastAPI services.
package observability

import (
	"fmt"
	"net/http"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Buckets matches the Python services' histogram buckets (seconds).
var Buckets = []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10}

type histogram struct {
	buckets []int64 // len(Buckets)+1, last is +Inf
	sum     float64
	count   int64
}

// Metrics is a thread-safe in-process registry.
type Metrics struct {
	mu        sync.Mutex
	service   string
	requests  map[string]int64 // key: route|method|status
	errors    map[string]int64 // key: route|method
	durations map[string]*histogram
}

// New returns a registry for the named service.
func New(service string) *Metrics {
	return &Metrics{
		service:   service,
		requests:  map[string]int64{},
		errors:    map[string]int64{},
		durations: map[string]*histogram{},
	}
}

type statusWriter struct {
	http.ResponseWriter
	status int
}

func (w *statusWriter) WriteHeader(code int) {
	w.status = code
	w.ResponseWriter.WriteHeader(code)
}

// Instrument wraps next, recording request count, latency, and 5xx errors.
// The Go 1.22+ ServeMux matched pattern (r.Pattern) is used as the
// low-cardinality route label; /metrics is excluded.
func (m *Metrics) Instrument(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/metrics" {
			next.ServeHTTP(w, r)
			return
		}
		start := time.Now()
		sw := &statusWriter{ResponseWriter: w, status: http.StatusOK}
		next.ServeHTTP(sw, r)
		elapsed := time.Since(start).Seconds()

		route := r.Pattern
		if route == "" {
			route = r.Method + " " + r.URL.Path
		}
		m.mu.Lock()
		m.requests[route+"|"+r.Method+"|"+strconv.Itoa(sw.status)]++
		h, ok := m.durations[route+"|"+r.Method]
		if !ok {
			h = &histogram{buckets: make([]int64, len(Buckets)+1)}
			m.durations[route+"|"+r.Method] = h
		}
		for i, b := range Buckets {
			if elapsed <= b {
				h.buckets[i]++
			}
		}
		h.buckets[len(Buckets)]++
		h.sum += elapsed
		h.count++
		if sw.status >= 500 {
			m.errors[route+"|"+r.Method]++
		}
		m.mu.Unlock()
	})
}

// Handler serves the Prometheus text exposition at /metrics.
func (m *Metrics) Handler() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "text/plain; version=0.0.4; charset=utf-8")
		fmt.Fprint(w, m.Render())
	})
}

func num(v float64) string {
	if v == float64(int64(v)) {
		return strconv.FormatInt(int64(v), 10)
	}
	return strconv.FormatFloat(v, 'f', -1, 64)
}

func labels(pairs ...string) string {
	if len(pairs) == 0 {
		return ""
	}
	var b strings.Builder
	b.WriteByte('{')
	for i := 0; i < len(pairs); i += 2 {
		if i > 0 {
			b.WriteByte(',')
		}
		fmt.Fprintf(&b, "%s=%q", pairs[i], pairs[i+1])
	}
	b.WriteByte('}')
	return b.String()
}

// Render produces the text exposition format (deterministic ordering).
func (m *Metrics) Render() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	var b strings.Builder
	b.WriteString("# HELP service_info Static service identity gauge (always 1).\n")
	b.WriteString("# TYPE service_info gauge\n")
	fmt.Fprintf(&b, "service_info%s 1\n", labels("service", m.service, "version", "0.1.0"))

	b.WriteString("# HELP http_requests_total Total HTTP requests by route/method/status.\n")
	b.WriteString("# TYPE http_requests_total counter\n")
	reqKeys := make([]string, 0, len(m.requests))
	for k := range m.requests {
		reqKeys = append(reqKeys, k)
	}
	sort.Strings(reqKeys)
	for _, k := range reqKeys {
		parts := strings.Split(k, "|")
		fmt.Fprintf(&b, "http_requests_total%s %d\n",
			labels("method", parts[1], "route", parts[0], "service", m.service, "status", parts[2]),
			m.requests[k])
	}

	b.WriteString("# HELP http_errors_total HTTP 5xx responses by route/method.\n")
	b.WriteString("# TYPE http_errors_total counter\n")
	errKeys := make([]string, 0, len(m.errors))
	for k := range m.errors {
		errKeys = append(errKeys, k)
	}
	sort.Strings(errKeys)
	for _, k := range errKeys {
		parts := strings.Split(k, "|")
		fmt.Fprintf(&b, "http_errors_total%s %d\n",
			labels("method", parts[1], "route", parts[0], "service", m.service), m.errors[k])
	}

	b.WriteString("# HELP http_request_duration_seconds Request latency histogram.\n")
	b.WriteString("# TYPE http_request_duration_seconds histogram\n")
	hKeys := make([]string, 0, len(m.durations))
	for k := range m.durations {
		hKeys = append(hKeys, k)
	}
	sort.Strings(hKeys)
	for _, k := range hKeys {
		parts := strings.Split(k, "|")
		h := m.durations[k]
		base := []string{"method", parts[1], "route", parts[0], "service", m.service}
		var cumulative int64
		for i, bound := range Buckets {
			cumulative += h.buckets[i]
			fmt.Fprintf(&b, "http_request_duration_seconds_bucket%s %d\n",
				labels(append(base, "le", num(bound))...), cumulative)
		}
		cumulative += h.buckets[len(Buckets)]
		fmt.Fprintf(&b, "http_request_duration_seconds_bucket%s %d\n",
			labels(append(base, "le", "+Inf")...), cumulative)
		fmt.Fprintf(&b, "http_request_duration_seconds_sum%s %s\n", labels(base...), strconv.FormatFloat(h.sum, 'f', 6, 64))
		fmt.Fprintf(&b, "http_request_duration_seconds_count%s %d\n", labels(base...), h.count)
	}
	return b.String()
}

File: services/_shared/observability.py
"""Lightweight, dependency-optional observability instrumentation (Stage 7.C).

``instrument_fastapi(app, service_name)`` wires three things onto any SOS
FastAPI service with a single call:

1. ``/metrics`` — Prometheus text-exposition endpoint backed by a tiny
   zero-dependency local registry (per-app instance, so tests are
   deterministic). Exposes ``http_requests_total``,
   ``http_request_duration_seconds`` (histogram) and ``http_errors_total``
   labelled by service/route/method, plus ``service_info``.
2. Optional OpenTelemetry wiring — only when the ``opentelemetry-*``
   packages are installed *and* ``OTEL_EXPORTER_OTLP_ENDPOINT`` is set.
   Every step is import-guarded and fail-soft: any missing package or
   exporter error falls back to the local registry only.
3. ``X-Request-ID`` propagation middleware — echoes an inbound request ID
   or mints a new one, exposes it on ``request.state.request_id`` and sets
   the response header.

Import convention (same as ``_shared.hashchain``): services insert the
``services/`` directory into ``sys.path`` before importing. Container
images that only ship ``app/`` guard the import and skip instrumentation.
"""
from __future__ import annotations

import os
import time
import uuid
from threading import Lock
from typing import Dict, Iterable, Optional, Tuple

HISTOGRAM_BUCKETS: Tuple[float, ...] = (
    0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
)

_LABEL_BLACKLIST_PATHS = frozenset({"/metrics"})


class LocalRegistry:
    """Minimal in-process metric registry (counters + fixed-bucket histograms).

    Pure stdlib, thread-safe, and entirely deterministic — no wall-clock or
    global state leaks between app instances because each ``instrument_fastapi``
    call creates its own registry.
    """

    def __init__(self) -> None:
        self._lock = Lock()
        # {(metric, labels-tuple): value}
        self._counters: Dict[Tuple[str, Tuple[Tuple[str, str], ...]], float] = {}
        # {(metric, labels-tuple): [bucket_counts..., +Inf], sum, count}
        self._histograms: Dict[
            Tuple[str, Tuple[Tuple[str, str], ...]],
            list,
        ] = {}

    @staticmethod
    def _key(metric: str, labels: Optional[Dict[str, str]]) -> Tuple[str, Tuple[Tuple[str, str], ...]]:
        return metric, tuple(sorted((labels or {}).items()))

    def inc(self, metric: str, labels: Optional[Dict[str, str]] = None, value: float = 1.0) -> None:
        with self._lock:
            key = self._key(metric, labels)
            self._counters[key] = self._counters.get(key, 0.0) + value

    def observe(self, metric: str, value: float, labels: Optional[Dict[str, str]] = None) -> None:
        with self._lock:
            key = self._key(metric, labels)
            entry = self._histograms.get(key)
            if entry is None:
                entry = [[0] * (len(HISTOGRAM_BUCKETS) + 1), 0.0, 0]
                self._histograms[key] = entry
            buckets, total, count = entry
            for i, bound in enumerate(HISTOGRAM_BUCKETS):
                if value <= bound:
                    buckets[i] += 1
            buckets[-1] += 1  # +Inf
            entry[1] = total + value
            entry[2] = count + 1

    # -- exposition -----------------------------------------------------

    @staticmethod
    def _fmt_labels(labels: Tuple[Tuple[str, str], ...], extra: Optional[Dict[str, str]] = None) -> str:
        merged = list(labels)
        if extra:
            merged.extend(sorted(extra.items()))
        if not merged:
            return ""
        inner = ",".join(f'{k}="{v}"' for k, v in merged)
        return "{" + inner + "}"

    @staticmethod
    def _num(value: float) -> str:
        return str(int(value)) if float(value).is_integer() else repr(value)

    def render_prometheus(self, service_name: str) -> str:
        """Render the registry in Prometheus text exposition format."""
        lines = [
            "# HELP service_info Static service identity gauge (always 1).",
            "# TYPE service_info gauge",
            f'service_info{{service="{service_name}",version="0.1.0"}} 1',
            "# HELP http_requests_total Total HTTP requests by route/method/status.",
            "# TYPE http_requests_total counter",
        ]
        with self._lock:
            counters = dict(self._counters)
            histograms = {k: [list(v[0]), v[1], v[2]] for k, v in self._histograms.items()}
        for (metric, labels), value in sorted(counters.items()):
            if metric.startswith("http_"):
                lines.append(f"{metric}{self._fmt_labels(labels)} {self._num(value)}")
        lines += [
            "# HELP http_errors_total HTTP 5xx responses by route/method.",
            "# TYPE http_errors_total counter",
            "# HELP http_request_duration_seconds Request latency histogram.",
            "# TYPE http_request_duration_seconds histogram",
        ]
        for (metric, labels), (buckets, total, count) in sorted(histograms.items()):
            cumulative = 0
            for i, bound in enumerate(HISTOGRAM_BUCKETS):
                cumulative += buckets[i]
                lines.append(
                    f"{metric}_bucket{self._fmt_labels(labels, {'le': self._num(bound)})} {cumulative}"
                )
            cumulative += buckets[-1]
            lines.append(
                f'{metric}_bucket{self._fmt_labels(labels, {"le": "+Inf"})} {cumulative}'
            )
            lines.append(f"{metric}_sum{self._fmt_labels(labels)} {self._num(round(total, 6))}")
            lines.append(f"{metric}_count{self._fmt_labels(labels)} {count}")
        return "\n".join(lines) + "\n"


def _route_label(request) -> str:
    """Best-effort low-cardinality route label (path template if routed)."""
    route = request.scope.get("route")
    if route is not None and getattr(route, "path", None):
        return route.path
    return request.url.path


def _try_setup_otel(app, service_name: str) -> bool:
    """Wire OpenTelemetry tracing+metrics when available and configured.

    Fail-soft by design: missing packages, unset endpoint, or any exporter
    error returns False and the local registry remains the only sink.
    """
    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "").strip()
    if not endpoint:
        return False
    try:
        from opentelemetry import metrics as otel_metrics, trace
        from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
        from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
        from opentelemetry.sdk.metrics import MeterProvider
        from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
    except ImportError:
        return False
    try:
        resource = Resource.create({"service.name": service_name})
        tracer_provider = TracerProvider(resource=resource)
        tracer_provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint)))
        trace.set_tracer_provider(tracer_provider)
        reader = PeriodicExportingMetricReader(OTLPMetricExporter(endpoint=endpoint))
        otel_metrics.set_meter_provider(MeterProvider(resource=resource, metric_readers=[reader]))
        FastAPIInstrumentor.instrument_app(app)
        return True
    except Exception:
        return False


def instrument_fastapi(app, service_name: str, registry: Optional[LocalRegistry] = None) -> None:
    """Attach /metrics, request-ID propagation, and optional OTel to ``app``.

    Idempotent per app instance: repeated calls with an explicit ``registry``
    re-use it; otherwise a fresh registry is created once and stored on
    ``app.state.observability_registry``.
    """
    from starlette.requests import Request
    from starlette.responses import PlainTextResponse

    registry = registry or getattr(app.state, "observability_registry", None) or LocalRegistry()
    app.state.observability_registry = registry
    app.state.otel_enabled = _try_setup_otel(app, service_name)

    @app.middleware("http")
    async def request_id_middleware(request: Request, call_next):
        request_id = request.headers.get("x-request-id") or uuid.uuid4().hex
        request.state.request_id = request_id
        response = await call_next(request)
        response.headers["X-Request-ID"] = request_id
        return response

    @app.middleware("http")
    async def metrics_middleware(request: Request, call_next):
        if request.url.path in _LABEL_BLACKLIST_PATHS:
            return await call_next(request)
        start = time.perf_counter()
        response = await call_next(request)
        elapsed = time.perf_counter() - start
        labels = {
            "service": service_name,
            "route": _route_label(request),
            "method": request.method,
            "status": str(response.status_code),
        }
        registry.inc("http_requests_total", labels)
        registry.observe(
            "http_request_duration_seconds",
            elapsed,
            {k: v for k, v in labels.items() if k != "status"},
        )
        if response.status_code >= 500:
            registry.inc(
                "http_errors_total",
                {k: v for k, v in labels.items() if k != "status"},
            )
        return response

    @app.get("/metrics", include_in_schema=False)
    def metrics_endpoint() -> PlainTextResponse:
        return PlainTextResponse(
            registry.render_prometheus(service_name),
            media_type="text/plain; version=0.0.4; charset=utf-8",
        )


def exposition_lines(registry: LocalRegistry, service_name: str) -> Iterable[str]:
    """Test helper: iterate rendered exposition lines."""
    return iter(registry.render_prometheus(service_name).splitlines())
